File: android/src/main/java/fr/greweb/rnwebglviewshot/RNWebGLTextureViewLoader.java

package fr.greweb.rnwebglviewshot;

import fr.greweb.rnwebgl.RNWebGLTextureCompletionBlock;
import fr.greweb.rnwebgl.RNWebGLTextureConfigLoader;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.uimanager.UIManagerModule;

public class RNWebGLTextureViewLoader extends ReactContextBaseJavaModule  implements RNWebGLTextureConfigLoader {

    private final ReactApplicationContext reactContext;

    public RNWebGLTextureViewLoader(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public boolean canLoadConfig(ReadableMap config) {
        return config.hasKey("view");
    }

    @Override
    public void loadWithConfig(ReadableMap config, RNWebGLTextureCompletionBlock callback) {
        int tag = config.getInt("view");
        try {
            UIManagerModule uiManager = this.reactContext.getNativeModule(UIManagerModule.class);
            uiManager.addUIBlock(new RNWebGLTextureViewUIBlock(config, tag, callback));
        }
        catch (Exception e) {
            callback.call(e, null);
        }
    }

    @Override
    public String getName() {
        return "RNWebGLTextureViewLoader";
    }
}

File: ios/RNWebGLTextureViewLoader.m
#import <React/RCTConvert.h>
#import <React/RCTImageSource.h>
#import <React/RCTBridge.h>
#import <React/RCTUIManager.h>
#if __has_include(<React/RCTUIManagerUtils.h>)
#import <React/RCTUIManagerUtils.h>
#endif
#import "RNWebGLTextureViewLoader.h"
#import "RNWebGLTextureView.h"

@implementation RNWebGLTextureViewLoader

RCT_EXPORT_MODULE()

@synthesize bridge = _bridge;

- (BOOL)canLoadConfig:(NSDictionary *)config {
  return [config objectForKey:@"view"] != nil;
}

- (void)loadWithConfig:(NSDictionary *)config
   withCompletionBlock:(RNWebGLTextureCompletionBlock)callback {
  dispatch_async(RCTGetUIManagerQueue(), ^{
    NSNumber *viewTag = [RCTConvert NSNumber:[config objectForKey:@"view"]];
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, UIView *> *viewRegistry) {
      UIView *view = viewRegistry[viewTag];
      if (!view) {
        callback([NSError errorWithDomain:@"RNWebGLViewShot" code:1 userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"No view found with reactTag: %@", viewTag] }], nil);
      }
      else {
        RNWebGLTextureView *obj = [[RNWebGLTextureView alloc] initWithConfig:config withView:view];
        callback(nil, obj);
      }
    }];
  });
}

@end
